physical/postgresql.go

```diff
@@ -175,3 +199,143 @@ func (m *PostgreSQLBackend) List(prefix string) ([]string, error) {
 
 	return keys, nil
 }
+
+func (m *PostgreSQLBackend) HAEnabled() bool {
+	return true
+}
+
+// PostgreSQLLock implements the Lock interface for PostgreSQL
+type PostgreSQLLock struct {
+	client *sql.DB
+	table  string
+	leader chan struct{}
+
+	key     string
+	value   string
+	vaultID string
+}
+
+// LockWith initializes a Postgres backend lock
+func (m *PostgreSQLBackend) LockWith(key, value string) (Lock, error) {
+	id, err := uuid.GenerateUUID()
+	if err != nil {
+		return nil, err
+	}
+	return &PostgreSQLLock{
+		client:  m.client,
+		table:   m.lock_table,
+		leader:  make(chan struct{}),
+		key:     key,
+		value:   value,
+		vaultID: id,
+	}, nil
+}
+
+// Lock grabs a lock, or waits until it is available
+func (m *PostgreSQLLock) Lock(stopCh <-chan struct{}) (<-chan struct{}, error) {
+	ctx, cancel := context.WithCancel(context.Background())
+	go func() {
+		<-stopCh
+		cancel()
+	}()
+	ticker := time.NewTicker(PostgreSQLLockRetryInterval)
+	defer ticker.Stop()
+	for {
+		_, err := m.client.ExecContext(
+			ctx,
+			"DELETE FROM "+m.table+" WHERE expiration < now()",
+		)
+		if err != nil {
+			return nil, err
+		}
```

Comment:
Sorry, I'm having a hard time following that transaction. What exactly is the `INSERT` that still happens? Are `$2` and `$3` the same or different? Can you explain what the `NOT IN` clause is trying to do (the double negation of `NOT IN` and `vault_id !=` is confusing me)?

With `SERIALIZABLE` transactions (in both postgres and CRDB), it's fairly simple to do this as a multi-step transaction:

    BEGIN;
    -- SERIALIZABLE is the default on CRDB but not on postgres
    SET TRANSACTION ISOLATION SERIALIZABLE;
    SELECT MAX(lock_lease_end) FROM vault_lock;
    -- if MAX(lock_lease_end) > NOW(), exit
    UPDATE vault_lock SET lock_lease_end = lock_lease_start + $1::INTERVAL WHERE vault_id=$2;
    COMMIT;

The performance of this transaction won't be great (it scans the entire vault_lock table since there is no index, and in both databases it may need to be retried in a loop to ensure serializability), but for occasional lock acquisition that's probably fine. Rewriting it as an UPDATE with a subquery can improve performance significantly (at least in CRDB) if you find that to be worth the complexity. 

Note that if there were an index on `lock_lease_end`, the `SELECT` statement in my transaction would not touch all the rows, just one row plus the gap in the index where any new lease would have to be inserted. In a serializable transaction, that's enough to ensure consistency.

Finally, note that `NOW()` can move backwards (in any database you may experience negative NTP adjustments; in CockroachDB you can additionally find that different nodes have slightly different clocks), so you need to be careful whenever you use `NOW()` that you also watch for times already in the database that may be in your "future".